### PR TITLE
[SPARK-34659] Fix that Web UI always correctly get appId

### PR DESCRIPTION
**Pull request by ornew**
Web UI does not correctly get appId when it has `proxy` or `history` in URL.

In my case, it happens on `https://jupyterhub.hosted.us/my-name/proxy/4040/executors/`.
Web developer console says: `jquery-3.4.1.min.js:2 GET https://jupyterhub.hosted.us/user/my-name/proxy/4040/api/v1/applications/4040/allexecutors 404`, and it shows blank pages to me.

There is relative issue in jupyterhub https://github.com/jupyterhub/jupyter-server-proxy/issues/57

https://github.com/apache/spark/blob/2526fdea481b1777b2c4a2242254b72b5c49d820/core/src/main/resources/org/apache/spark/ui/static/utils.js#L93-L105

It should not get from document.baseURI.
A request will occur, but performance impacts will be a bit.

### What changes were proposed in this pull request?

This always get an appId using the API.

### Why are the changes needed?

The UI does not appear correctly in some environments. For example, there is a Jupyterhub.

### Does this PR introduce _any_ user-facing change?

No, this is bug fix.

### How was this patch tested?

I see it correctly works in my browser.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @ornew .

cc @gengliangwang

**Comment by gengliangwang** [comment]
@ornew could you show reproduce steps on Spark? It seems that the jupyterhub is using a different Spark UI URL from Spark.
> Also the code change breaks the logic. Spark can get the app id without access rest API if the URL contains proxy or history.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @ornew .

**Comment by ornew** [comment]
@dongjoon-hyun @gengliangwang Thank you for your reply.

> @ornew could you show reproduce steps on Spark? It seems that the jupyterhub is using a different Spark UI URL from Spark.

It's easy. Please run JupyterHub with [jupyter-server-proxy](https://github.com/jupyterhub/jupyter-server-proxy), and PySpark.

I build Jupyter Hub on Kubernetes and provide a sandbox for a large number of users. Accessing the Spark UI requires some kind of proxy. JupyterHub's Server Proxy allows you to access ports in your sandbox environment without user interaction.

```python
from pyspark import *
from pyspark.sql import *

spark = SparkSession.builder.getOrCreate()
```

スクリーンショット 2021-03-16 14 58 57

When accessing the Spark UI by port, the Jupyter Server Proxy path contains `proxy`, which causes incorrect parsing.

スクリーンショット 2021-03-16 15 01 56

**Comment by gengliangwang** [comment]
@ornew Is it possible to fix it in JupyterHub? The issue is not on Spark itself.

**Comment by ornew** [comment]
@gengliangwang I think this is a Spark issue. 

This always happens when the path contains `proxy` or `history` if without JupyterHub also. There are many use cases where the path does not include the `appId`, such as standalone or Kubernetes. Also there are many opportunities to access it via a proxy. Would you please reconsider about and the logic of getting appId that depends on the environment and URL?

**Comment by gengliangwang** [comment]
@ornew I mean, if we can't reproduce the issue on a Spark cluster, then it is not the issue of Spark itself.
Spark does support running behind a reverse proxy, see https://github.com/apache/spark/pull/29820 for details.

File: core/src/main/resources/org/apache/spark/ui/static/utils.js [hunk 1]
```diff
@@ -90,26 +90,27 @@ function formatLogsCells(execLogs, type) {
...
-  }
   // Looks like Web UI is running in standalone mode
   // Let's get application-id using REST End Point
   $.getJSON(uiRoot + "/api/v1/applications", function(response, status, jqXHR) {
     if (response && response.length > 0) {
       var appId = response[0].id;
       cb(appId);
       return;
+    } else {
+      var words = document.baseURI.split('/');
```
Comment (gengliangwang) [inline core/src/main/resources/org/apache/spark/ui/static/utils.js:101]: BTW, have you tested the code changes for Spark UI behind proxy and Spark UI of History server?

**Comment by PerilousApricot** [comment]
@gengliangwang -- I actually have a very simple reproducer using nginx as a reverse proxy and not jupyterhub (to eliminate that failure mode). The following script will set up the proxy, note that it redirects `/user/PerilousApricot/proxy/4040/` to the root of the spark webUI 

**proxy-fail.sh**
```bash
#!/bin/bash


cat << \EOT > nginx.conf
user  nginx;
worker_processes  auto;
error_log  /var/log/nginx/error.log notice;
pid        /var/run/nginx.pid;
events {
    worker_connections  1024;
}
http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;
    log_format  main  '[$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_x_forwarded_for"';
    access_log  /dev/stdout  main;
    server {
        listen       5050;
        server_name  localhost;
        location /user/PerilousApricot/proxy/4040/ {
            error_log  /dev/stderr debug;
            proxy_pass http://localhost:4040/;
            proxy_pass_header Content-Type;
        }        
    }
}

EOT

docker run -it --rm=true --name spark-31174-proxy --network=host -v $(pwd)/nginx.conf:/etc/nginx/nginx.conf:ro nginx
```

Run that proxy in one terminal, then run pyspark:
```
SPARK_PUBLIC_DNS=localhost:5050/user/PerilousApricot/proxy/4040/jobs/ pyspark --conf spark.ui.reverseProxyUrl=http://localhost:5000/user/PerilousApricot/proxy/4040/ --conf spark.driver.extraJavaOptions="-Dlog4j.debug=true" --conf spark.ui.proxyBase=/user/PerilousApricot/proxy/4040/ --conf spark.app.name=proxyApp
```


Open `http://localhost:5050/user/PerilousApricot/proxy/4040/executors/` in a browser with "developer mode" enabled to watch the traffic come by. You will see a number of successful requests to various resources like:

```
 http://localhost:5050/user/PerilousApricot/proxy/4040//static/webui.css
 http://localhost:5050/user/PerilousApricot/proxy/4040//static/webui.js
 ```

Notice, however that there is a failed request (and the reason of this PR) - 
```
http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/4040/allexecutors
```

If you run curl manually on both that URL, you can see that it fails both at the reverse proxy and at the actual webui itself:
```
curl -v -o /dev/null http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/4040/allexecutors
curl -v -o /dev/null http://localhost:4040/api/v1/applications/4040/allexecutors
```

But if you copy-paste the appId from the spark console (in my case I have: `Spark context available as 'sc' (master = local[*], app id = local-1639522961946).`), the following two requests succeed:
```
curl http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/local-1639522961946
curl -v -o /dev/null http://localhost:4040/api/v1/applications/local-1639522961946
```

To confirm the issue, let's restart the proxy and pyspark, but instead of proxying `/user/PerilousApricot/proxy/4040/`, let's instead proxy to `/user/PerilousApricot/yxorp/4040/` (note that there is no "proxy" in the proxied URL). First execute
**proxy-win.sh**
```
#!/bin/bash


cat << \EOT > nginx.conf
user  nginx;
worker_processes  auto;
error_log  /var/log/nginx/error.log notice;
pid        /var/run/nginx.pid;
events {
    worker_connections  1024;
}
http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;
    log_format  main  '[$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_x_forwarded_for"';
    access_log  /dev/stdout  main;
    server {
        listen       5050;
        server_name  localhost;
        location /user/PerilousApricot/yxorp/4040/ {
            #error_log  /dev/stderr debug;
            proxy_pass http://localhost:4040/;
            #proxy_redirect     off;
            proxy_pass_header Content-Type;
            #rewrite /user/PerilousApricot/yxorp/4040(/.*|$) $1  break;
        }        
    }
}

EOT

docker run -it --rm=true --name spark-31174-proxy --network=host -v $(pwd)/nginx.conf:/etc/nginx/nginx.conf:ro nginx
```

and then run in a different terminal
```
SPARK_PUBLIC_DNS=localhost:5050/user/PerilousApricot/yxorp/4040/jobs/ pyspark --conf spark.ui.reverseProxyUrl=http://localhost:5000/user/PerilousApricot/yxorp//4040/ --conf spark.driver.extraJavaOptions="-Dlog4j.debug=true" --conf spark.ui.proxyBase=/user/PerilousApricot/yxorp/4040/ --conf spark.app.name=proxyApp
```

Open `http://localhost:5050/user/PerilousApricot/yxorp/4040//executors/` and you can see that the page renders properly. Looking at the development console, you will see that instead of attempting to open

```
http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/4040/allexecutors
```

this version requests the status of the executors from
```
http://localhost:5050/user/PerilousApricot/yxorp/4040//api/v1/applications/local-1639523380430/allexecutors
```

I hope this is enough to show that @ornew did the right analysis -- Th fault isn't with jupyterhub, it is simply the fact that the logic that tries to look up the appId chokes if there is a path element named "proxy" in the URL.

Can you please re-examine this?

**Comment by gengliangwang** [comment]
@PerilousApricot  Are you running Spark as a cluster? If yes, Spark supports reverse proxy, see the following PRs for details:
https://github.com/apache/spark/pull/13950/
https://github.com/apache/spark/pull/29820

**Comment by PerilousApricot** [comment]
Hi @gengliangwang yes, I am.

Thanks for the pointers, but the point of the PR is that there is a bug in how the reverse proxying is handled. If you see the reproducer, I am using the config options mentioned in #13950 and #29820.

**Comment by PerilousApricot** [comment]
In the current master, when you reverse proxy to
```
http://localhost:5050/user/PerilousApricot/proxy/4040/
```
then Spark UI tries to do an API call to
```
http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/4040/allexecutors
```
to retrieve the executor status, but this is incorrect, it should be
```
http://localhost:5050/user/PerilousApricot/proxy/4040/api/v1/applications/local-1639523380430/allexecutors
```
(where local-1639523380430 is the appId of the SparkContext).

The problem is that Spark itself bungles the handling of the appId. You said in https://github.com/apache/spark/pull/31774#issuecomment-802997989 that the problem was unreproducible in a Spark cluster, I hope that the reproducer I put in the comment above is enough to show the issue. Please let me know if I can help clarify it better.

**Comment by PerilousApricot** [comment]
Hello @gengliangwang and Happy New years! I'm back from vacation and was wondering if you had further thoughts on this issue. Were you able to reproduce the bug?

**Comment by PerilousApricot** [comment]
Hello @gengliangwang checking up on this issue. I gave a reproducer above that clearly shows the issue. Have you had a chance to take a look?

**Comment by gengliangwang** [comment]
@PerilousApricot I will take a close look before the 3.3 release.

**Comment by PerilousApricot** [comment]
@gengliangwang Thank you very much! This would be a huge relief for our use-case

**Comment by PerilousApricot** [comment]
This is really important for us, so I hope it doesn't slip through the
cracks. Is there anyone else available to review?--
It's dark in this basement.

**Comment by dongjoon-hyun** [comment]
According to the recent comments, I removed `Stale` tag and reopened this. However, the PR itself seems to have conflicts already.

**Comment by PerilousApricot** [comment]
Hello @dongjoon-hyun I'm happy to update the PR if there is someone available to review and merge the resulting code.

**Comment by MaxGekk** [comment]
@gengliangwang @dongjoon-hyun Could you, please, help to review this if we plan to have it in 3.3 (just in case, it is in the allow list).

**Comment by MaxGekk** [comment]
@ornew Please, resolve conflicts.

**Comment by gengliangwang** [comment]
> SPARK_PUBLIC_DNS=localhost:5050/user/PerilousApricot/proxy/4040/jobs/ pyspark --conf spark.ui.reverseProxyUrl=http://localhost:5000/user/PerilousApricot/proxy/4040/ --conf spark.driver.extraJavaOptions="-Dlog4j.debug=true" --conf spark.ui.proxyBase=/user/PerilousApricot/proxy/4040/ --conf spark.app.name=proxyApp

@PerilousApricot should all the port be 5050? You are setting the revert proxy URL as `http://localhost:5000`, which is confusing.

**Comment by PerilousApricot** [comment]
@MaxGekk there are some things @gengliangwang and @dongjoon-hyun would like fixed with this PR, but I'm hopeful that we can converge soon on a solution they'll accept.

**Comment by lordk911** [comment]
@PerilousApricot 
https://github.com/jupyterhub/jupyter-server-proxy/issues/57#issuecomment-2918044837

use this to workaround

**Pull request closed without merging**